File: lib/core/utils/geohash.dart
/// Minimal geohash implementation for proximity queries
/// (geofire-style prefix ranges on Firestore string fields).
library;

const String _base32 = '0123456789bcdefghjkmnpqrstuvwxyz';

/// Encodes a coordinate to a geohash of [precision] characters.
String geohashEncode(double latitude, double longitude, {int precision = 9}) {
  var latLo = -90.0, latHi = 90.0;
  var lonLo = -180.0, lonHi = 180.0;
  final buffer = StringBuffer();
  var isEven = true;
  var bit = 0;
  var ch = 0;

  while (buffer.length < precision) {
    if (isEven) {
      final mid = (lonLo + lonHi) / 2;
      if (longitude >= mid) {
        ch = (ch << 1) + 1;
        lonLo = mid;
      } else {
        ch = ch << 1;
        lonHi = mid;
      }
    } else {
      final mid = (latLo + latHi) / 2;
      if (latitude >= mid) {
        ch = (ch << 1) + 1;
        latLo = mid;
      } else {
        ch = ch << 1;
        latHi = mid;
      }
    }
    isEven = !isEven;
    if (++bit == 5) {
      buffer.write(_base32[ch]);
      bit = 0;
      ch = 0;
    }
  }
  return buffer.toString();
}

({double latLo, double latHi, double lonLo, double lonHi}) _bounds(
    String geohash) {
  var latLo = -90.0, latHi = 90.0;
  var lonLo = -180.0, lonHi = 180.0;
  var isEven = true;

  for (final char in geohash.split('')) {
    final cd = _base32.indexOf(char);
    assert(cd >= 0, 'Invalid geohash character: $char');
    for (var mask = 16; mask > 0; mask >>= 1) {
      if (isEven) {
        final mid = (lonLo + lonHi) / 2;
        if (cd & mask != 0) {
          lonLo = mid;
        } else {
          lonHi = mid;
        }
      } else {
        final mid = (latLo + latHi) / 2;
        if (cd & mask != 0) {
          latLo = mid;
        } else {
          latHi = mid;
        }
      }
      isEven = !isEven;
    }
  }
  return (latLo: latLo, latHi: latHi, lonLo: lonLo, lonHi: lonHi);
}

/// The cell itself plus its 8 neighbors, at the same precision.
/// Covers a radius of at least half a cell in every direction, which is
/// why callers pick the precision to match their search radius.
List<String> geohashCoverage(String geohash) {
  final b = _bounds(geohash);
  final latCenter = (b.latLo + b.latHi) / 2;
  final lonCenter = (b.lonLo + b.lonHi) / 2;
  final dLat = b.latHi - b.latLo;
  final dLon = b.lonHi - b.lonLo;

  final cells = <String>{};
  for (var dy = -1; dy <= 1; dy++) {
    for (var dx = -1; dx <= 1; dx++) {
      final lat = (latCenter + dy * dLat).clamp(-89.999999, 89.999999);
      var lon = lonCenter + dx * dLon;
      lon = ((lon + 180) % 360 + 360) % 360 - 180;
      cells.add(geohashEncode(lat, lon, precision: geohash.length));
    }
  }
  return cells.toList();
}

File: functions/src/geohash.ts
/**
 * Minimal geohash implementation, mirroring lib/core/utils/geohash.dart.
 */

const BASE32 = "0123456789bcdefghjkmnpqrstuvwxyz";

export function geohashEncode(
  latitude: number,
  longitude: number,
  precision = 9
): string {
  let latLo = -90, latHi = 90, lonLo = -180, lonHi = 180;
  let hash = "";
  let isEven = true;
  let bit = 0;
  let ch = 0;

  while (hash.length < precision) {
    if (isEven) {
      const mid = (lonLo + lonHi) / 2;
      if (longitude >= mid) {
        ch = (ch << 1) + 1;
        lonLo = mid;
      } else {
        ch = ch << 1;
        lonHi = mid;
      }
    } else {
      const mid = (latLo + latHi) / 2;
      if (latitude >= mid) {
        ch = (ch << 1) + 1;
        latLo = mid;
      } else {
        ch = ch << 1;
        latHi = mid;
      }
    }
    isEven = !isEven;
    if (++bit === 5) {
      hash += BASE32[ch];
      bit = 0;
      ch = 0;
    }
  }
  return hash;
}

function bounds(geohash: string): {
  latLo: number; latHi: number; lonLo: number; lonHi: number;
} {
  let latLo = -90, latHi = 90, lonLo = -180, lonHi = 180;
  let isEven = true;

  for (const char of geohash) {
    const cd = BASE32.indexOf(char);
    for (let mask = 16; mask > 0; mask >>= 1) {
      if (isEven) {
        const mid = (lonLo + lonHi) / 2;
        if (cd & mask) lonLo = mid;
        else lonHi = mid;
      } else {
        const mid = (latLo + latHi) / 2;
        if (cd & mask) latLo = mid;
        else latHi = mid;
      }
      isEven = !isEven;
    }
  }
  return {latLo, latHi, lonLo, lonHi};
}

/** The cell itself plus its 8 neighbors, at the same precision. */
export function geohashCoverage(geohash: string): string[] {
  const b = bounds(geohash);
  const latCenter = (b.latLo + b.latHi) / 2;
  const lonCenter = (b.lonLo + b.lonHi) / 2;
  const dLat = b.latHi - b.latLo;
  const dLon = b.lonHi - b.lonLo;

  const cells = new Set<string>();
  for (let dy = -1; dy <= 1; dy++) {
    for (let dx = -1; dx <= 1; dx++) {
      const lat = Math.max(-89.999999,
        Math.min(89.999999, latCenter + dy * dLat));
      let lon = lonCenter + dx * dLon;
      lon = ((lon + 180) % 360 + 360) % 360 - 180;
      cells.add(geohashEncode(lat, lon, geohash.length));
    }
  }
  return [...cells];
}

/** Haversine distance in kilometers. */
export function distanceKm(
  lat1: number, lon1: number, lat2: number, lon2: number
): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 6371 * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}
